### PR TITLE
Docs: reword dbm.gnu introduction

### PR DESCRIPTION
**Pull request by erlend-aasland**
Also:
- consistently spell GDBM as GDBM
- silence gdbm class refs
- improve accuracy of dbm.gdbm.open() spec

----
📚 Documentation preview 📚: https://cpython-previews--114548.org.readthedocs.build/

### CODE REVIEW

File: Doc/library/dbm.rst [hunk 1]
```diff
@@ -137,37 +137,39 @@ then prints out the contents of the database::
...
-The :mod:`dbm.gnu` module provides an interface to the GNU DBM library.
-``dbm.gnu.gdbm`` objects behave like mappings (dictionaries), except that keys and
-values are always converted to bytes before storing.  Printing a ``gdbm``
-object doesn't print the
-keys and values, and the :meth:`items` and :meth:`values` methods are not
-supported.
+:class:`!gdbm` objects behave like :term:`mappings <mapping>`,
+except that keys and values are always converted to :class:`bytes` before storing.
+Printing a :class:`!gdbm` object doesn't print the keys and values,
+and the :meth:`!items` and :meth:`!values` methods are not supported.
```
Comment (erlend-aasland) [inline Doc/library/dbm.rst]: Reduced duplication (consolidate the two first sentences in the two paragraphs), use a glossary link instead of parenthesised hint, and mark up stuff properly.

File: Doc/library/dbm.rst [hunk 2]
```diff
@@ -137,37 +137,39 @@ then prints out the contents of the database::
...
+
+.. note::
+   The file formats created by :mod:`dbm.gnu` and :mod:`dbm.ndbm` are
+   incompatible and can not be used interchangeably.
```
Comment (erlend-aasland) [inline Doc/library/dbm.rst:161]: IMO, this stands out better as a note instead of inlined. We should also mention this in the dbm.ndbm section. We can use a `|placeholder|` for that.

File: Doc/library/dbm.rst [hunk 3]
```diff
@@ -137,37 +137,39 @@ then prints out the contents of the database::
...
 
 .. exception:: error
 
    Raised on :mod:`dbm.gnu`-specific errors, such as I/O errors. :exc:`KeyError` is
    raised for general mapping errors like specifying an incorrect key.
 
 
-.. function:: open(filename[, flag[, mode]])
+.. function:: open(filename, flag="r", mode=0o666)
```
Comment (erlend-aasland) [inline Doc/library/dbm.rst]:
```suggestion
.. function:: open(filename, flag="r", mode=0o666, /)
```

File: Doc/library/dbm.rst (hunk 1 above)
Comment (ezio-melotti) [reply Doc/library/dbm.rst]: Not sure I like this paragraph too much.  It starts by saying that they behave like mappings, and then mentions a single exception.  Then it adds two more exceptions, saying what it doesn't print rather than what it prints.

I'd list list all the exceptions together, and mention what it prints (or maybe show it in an example somewhere).

Comment (erlend-aasland) [reply Doc/library/dbm.rst]: I'm not sure if the "print" subsentence is needed. It seems strange to me to even mention it.

Comment (erlend-aasland) [reply Doc/library/dbm.rst]: How about this:

```
:class:`!gdbm` objects behave like :term:`mappings <mapping>`,
except that keys and values are always converted to :class:`bytes` before storing,
and the :meth:`!items` and :meth:`!values` methods are not supported.
```

Comment (ezio-melotti) [reply Doc/library/dbm.rst]: That's definitely better!  Perhaps we could use "similarly to" instead of "like" (to me "like" has stronger implications).

Regarding the printed output, how do the `str`/`repr` look like?  If it's something worth mentioning, you could add a sentence like ``In addition, printing a :class:`!gdbm` shows a nicely formatted table.`` (or whatever).

Comment (erlend-aasland) [reply Doc/library/dbm.rst]: Yes, "similar to" is better than "like"; good suggestion! Regarding printed output, I'm not sure it is worth mentioning at all.

File: Doc/library/dbm.rst (hunk 2 above)
Comment (erlend-aasland) [reply Doc/library/dbm.rst:161]: I'll apply the placeholder in #114549.

File: Doc/library/dbm.rst (hunk 1 above)
Comment (erlend-aasland) [reply Doc/library/dbm.rst]: Applied!

File: Doc/library/dbm.rst (hunk 3 above)
Comment (ezio-melotti) [reply Doc/library/dbm.rst]: This conversation is marked as resolved, but the proposed change doesn't seem to have been applied.

Comment (erlend-aasland) [reply Doc/library/dbm.rst]: Hm, I thought I applied it locally; perhaps I did not git push.

Comment (erlend-aasland) [reply Doc/library/dbm.rst]: Aha, I see my confusion; I applied the change to the wrong line!

File: Doc/library/dbm.rst [hunk 4]
```diff
@@ -51,7 +51,7 @@ the Oracle Berkeley DB.
 .. |flag_n| replace::
    Always create a new, empty database, open for reading and writing.
 
-.. function:: open(file, flag='r', mode=0o666)
+.. function:: open(file, flag='r', mode=0o666, /)
```
Comment (erlend-aasland) [inline Doc/library/dbm.rst]: Revert this change.

```suggestion
.. function:: open(file, flag='r', mode=0o666)
```

**Review by hugovk — APPROVED** [review]
Thanks!

**Comment by erlend-aasland** [comment]
Thanks for the review, Hugo!
